### PR TITLE
charr_simulation_pipeline

### PR DESCRIPTION
**Pull request by wlu04**
I have three major concerns at this moment:
1. whether jobs named `j_index` are written in the correct way? They should take a few seconds to run on my laptop but have taken forever to run on batch. 
2. I tried to incorporate the `merge_gvcfs` in `run_vds_combiner.py` to the main `pipeline.py` script, but it always triggers an error before submitting to batch.
```
hailtop.httpx.ClientResponseError: 400, message=‘Bad Request’, url=URL(‘https://batch.hail.is/api/v1alpha/batches/create-fast’) body=‘400: jobs[6].resources.storage must match regex: [+]?((?:[0-9]*[.])?[0-9]+)([KMGTP][i]?)?’
```
3. I am not sure which files are necessary to checkpoint and check whether they exist before running.

### CODE REVIEW

File: batch_verifybamid.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    :return: True if the object exists
+    """
+    if path.startswith('gs://'):
+        bucket = path.replace('gs://', '').split('/')[0]
+        path = path.replace('gs://', '').split('/', maxsplit=1)[1]
+        path = path.rstrip('/')  # '.mt/' -> '.mt'
+        if any(path.endswith(f'.{suf}') for suf in ['mt', 'ht']):
+            path = os.path.join(path, '_SUCCESS')
+        gs = storage.Client()
+        return gs.get_bucket(bucket).get_blob(path)
```
Comment (konradjk) [inline batch_verifybamid.py]: this is fine, but you can also do `hl.hadoop_exists()`

File: batch_verifybamid.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        f"file://{local}/contam_temp.selfSM",
+    )
+    with open(f"{local}/contam_temp.selfSM", 'rt') as selfSM:
+        reader = csv.DictReader(selfSM, delimiter='\t')
+        i = 0
+        for row in reader:
+            if float(row["FREELK0"]) == 0 and float(row["FREELK1"]) == 0:
+                # a zero value for the likelihoods implies no data. This usually indicates a problem rather than a real event.
+                # if the bam isn't really empty, this is probably due to the use of a incompatible reference build between
+                # vcf and bam.
```
Comment (konradjk) [inline batch_verifybamid.py:52]: You could probably explicitly check the number of reads in the BAM

File: batch_verifybamid.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    #     )
+
+
+    ncpu = 8  # ~ 8G/core ~ 64G
+    # path, name = input_cram_path.rsplit('/', 1)
+    # output_prefix = name[:-5].split('.')[0]
+
+    j = b.new_job(f'Run_VerifyBamID_{output_prefix}')
+    j.image(IMAGE)
+    j.memory('highmem')
```
Comment (konradjk) [inline batch_verifybamid.py]: Does this need to be highmem?

File: batch_verifybamid.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    j.cpu(ncpu)
```
Comment (konradjk) [inline batch_verifybamid.py]: This is all fine if you prefer it this way but you can also chain these together (`j.image(IMAGE).memory(...).storage(...)`)

File: batch_verifybamid.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+                ref_fasta_path=REF_FASTA,
+                contamination_sites_path=CONTAM_SITES,
+                output_path=tmp_bucket,
+                disable_sanity_check=args.disable_sanity_check,
+            )
+
+            b.run()
+        else:
+            ht = hl.import_table('gs://gnomad-wenhan/verifybamid/gnomad_v3_hgdp_low_contam.txt').key_by('s')
+            cram_paths = ht.to_pandas().final_cram_path.tolist()
```
Comment (konradjk) [inline batch_verifybamid.py]: `ht.aggregate(hl.agg.collect(ht.final_cram_path))`

Comment (konradjk) [reply batch_verifybamid.py]: And if you wanted to get fancy you could do `cram_data = ht.aggregate(hl.struct(cram_path=hl.agg.collect(ht.final_cram_path), crai_path=...))` to do it all in one

File: batch_verifybamid.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+                    billing_project=BILLING_PROJECT,
+                    remote_tmpdir=tmp_bucket,
+                )
+
+                b = hb.Batch(
+                    'VerifyBamID',
+                    backend=backend,
+                )
+
+                for i in range(0, hl.eval(ht.count())):
```
Comment (konradjk) [inline batch_verifybamid.py]: you need neither the 0, nor the hl.eval here `range(ht.count())` will get the job done

Comment (konradjk) [reply batch_verifybamid.py]: but even better, `len(cram_paths)`

Comment (konradjk) [reply batch_verifybamid.py]: or best yet:
```
for cram, crai in zip(cram_paths, crai_paths):
    run_vbid(input_cram_path=cram_path, ...)
```

File: batch_verifybamid.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+    finally:
+        logger.info("Copying log to logging bucket...")
+        hl.copy_log(f"{tmp_bucket}log/")
+
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--input-cram-path', type=str)
```
Comment (konradjk) [inline batch_verifybamid.py]: type=str is default and not needed

File: run_vds_combiner.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        )
+        b = hb.Batch(
+            name=f"Merging-Files",
+            default_python_image='us-central1-docker.pkg.dev/broad-mpg-gnomad/wlu/hail/hail-pysam-samtools:latest',
+            backend=backend,
+        )
+
+        sample_ids = pd.read_csv(args.selected_samples, sep=',')
+        for i in range(len(sample_ids)):
+            gvcfs_to_merge = hl.utils.hadoop_ls(f'{args.input_gvcf_bucket}/{sample_ids[i]}/variant-calling/*.vcf')
```
Comment (konradjk) [inline run_vds_combiner.py]: they're uncompressed?

File: simulation_utils/cram_decontam_func.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+def get_var_dict_from_gvcf(input_gvcf: str):
+    # gvcf -> nested dictionary: {chromosome : { position : infomration ...}}
+    # os.system(f"ls {input_gvcf}")
+    print(pysam.__version__)
+    var_dict = {}
+    gvcf_in = VariantFile(input_gvcf)  # auto-detect input format
+    for rec in gvcf_in.fetch():
+        if (rec.samples[0]["GT"][0] != 0) | (rec.samples[0]["GT"][1] != 0):
+            if rec.contig not in var_dict.keys():
```
Comment (konradjk) [inline simulation_utils/cram_decontam_func.py:38]: This is fine if you'd like (although you don't have to call `.keys()`, rather `in var_dict` is sufficient), but you should check out `from collections import defaultdict`

File: simulation_utils/cram_decontam_func.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            var_dict[rec.contig][rec.pos] = (
+                rec.ref,
+                rec.alts[0],
+                rec.alts[1],
+                rec.samples[0]["GT"],
+            )
+    return var_dict
+
+def write_var_dict_from_gvcf(output: str, gvcf_dict: dict):
+    f = open(f"{output}", "wb")
```
Comment (konradjk) [inline simulation_utils/cram_decontam_func.py]:
```
with open(output, 'wb') as f:
    pickle.dump...
```

File: simulation_utils/cram_decontam_func.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    f.close()
+
+def read_pickle_dict(dict_path:str):
+    dict = pd.read_pickle(dict_path)
+    return dict
+
+def run_gvcf_dict(b: hb.batch, gvcf_file_name: str, gvcf_path: Tuple):
+    gvcf_dict_path = f"{MY_BUCKET}/gvcf_dicts/{gvcf_file_name}_gvcf_dict.pkl"
+    sample_id = gvcf_file_name.split('.')[0]
+    j = b.new_python_job(name=f"Run_gvcf_dict_{sample_id}")
```
Comment (konradjk) [inline simulation_utils/cram_decontam_func.py:63]: You might consider a fixed name, but setting job attributes `j.attributes({'sample_id': sample_id})` i think. You can do this throughout with all the jobs (for the mixing one, do `{'sample': s, 'contaminant': s2}` etc)

File: simulation_utils/cram_decontam_func.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    j._machine_type = "n1-highmem-16"
+    j.storage("500Gi")
+    if hl.hadoop_exists(gvcf_dict_path):
+        input_gvcf_dict = b.read_input(gvcf_dict_path)
+        gvcf_dict = j.call(read_pickle_dict, input_gvcf_dict)
+    else:
+        input_gvcf = b.read_input_group(**{"gvcf":gvcf_path[0], "index": f'{gvcf_path[1]}'})
```
Comment (konradjk) [inline simulation_utils/cram_decontam_func.py]: `gvcf=gvcf_path[0], index=gvcf_path[1]`

File: pipeline.py [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+                    gvcf_dict,
+                    input_cram_file,
+                    input_ref_fasta,
+                    j_cram.ofile,
+                    chromosome
+                )
+
+                logger.info(f'Indexing contamination free cram: {sample_id}-{chromosome}...')
+                j_index = b.new_job(name=f"Index_{sample_id}_{chromosome}_contam_free_file")
+                depend_on = j_index
```
Comment (konradjk) [inline pipeline.py]: legacy code?

Comment (wlu04) [reply pipeline.py]: > legacy code?

I don't think it is. but do you think this is unnecessary in some way?

File: batch_verifybamid.py (hunk 3 above)
Comment (wlu04) [reply batch_verifybamid.py]: > Does this need to be highmem?

Not sure about this, but I think I did put it here for some reason. Let me run a test and see.

**Comment by wlu04** [comment]
> I have three major concerns at this moment:
> 
> 1. whether jobs named `j_index` are written in the correct way? They should take a few seconds to run on my laptop but have taken forever to run on batch.
> 2. I tried to incorporate the `merge_gvcfs` in `run_vds_combiner.py` to the main `pipeline.py` script, but it always triggers an error before submitting to batch.
> 
> ```
> hailtop.httpx.ClientResponseError: 400, message=‘Bad Request’, url=URL(‘https://batch.hail.is/api/v1alpha/batches/create-fast’) body=‘400: jobs[6].resources.storage must match regex: [+]?((?:[0-9]*[.])?[0-9]+)([KMGTP][i]?)?’
> ```
> 
> 3. I am not sure which files are necessary to checkpoint and check whether they exist before running.

1 & 2 have been fixed. 1 is due to the invalid image for samtools command, and 2 is due to the misspecification of storage. still running some tests to figure out 3.

File: run_vds_combiner.py (hunk 8 above)
Comment (wlu04) [reply run_vds_combiner.py:22]: Yeah, the files from haplotype caller were suffixed with `.vcf`. I think there would be an error if we use `vcf.gz` and then merge them. but I can't really remember what that was.

Comment (wlu04) [reply run_vds_combiner.py]: > they're uncompressed?

Yeah, the files from haplotype caller were suffixed with .vcf. I think there would be an error if we use vcf.gz and then merge them. but I can't really remember what that was.

**Pull request merged**